accepted/000070-cluster_awareness.md

```diff
@@ -0,0 +1,321 @@
+- Managing an existing cluster has priority over deploying a new cluster from the Uyuni/SUSE Manager perspective
+- Deploying and provisioning the underlying infrastructure to host the cluster product is not covered in this RFC: for this purpose, consult [SUMA as a Virtualization Solution with Multi-Tenancy](https://confluence.suse.com/pages/viewpage.action?pageId=431620395)
+- We decided not to install the software management stack on the Uyuni/SUSE Manager itself but rather to have a dedicated management node for each cluster to avoid any kind of conflicts. This requirement may be relaxed in the future and Uyuni/SUSE Manager may be the management node of any cluster.
+- SES 6 requires to have a salt-master as part of the cluster - this will clash with the idea above of Uyuni/SUSE Manager being the salt-master of the minion. SES 7 is still under development but expected to be released in late 2020 and will probably relax this requirement. Uyuni/SUSE Manager will target SES 7: we consulted the available documentation for SES 7 (see References) and for everything that was not clear at the time of writing we looked at the documentation for SES 6.
+- This is an MVP: only basic functionality for a cluster will be covered in this first iteration
+
+## A new Cluster object
+
+A new “Clusters” item in the main menu will be added.
+The "Clusters" section will be available to the Administrative Roles and the new "Cluster Administrator" role.
+
+Under Clusters, there will be:
+
+- An “Import existing cluster” item that allows importing an existing cluster
+- A “Create new cluster” item that allows to deploy and define a new cluster in Uyuni/SUSE Manager
+- An “Overview” item that lists all the clusters currently known to Uyuni/SUSE Manager
+
+A cluster is identified in Uyuni/SUSE Manager by:
+- the cluster name (and an associated cluster secret)
+- the cluster type
+- the management node (hostname) - for clusters that do not requires a management node (e.g. SHAP) one node of the cluster will be used by Uyuni/SUSE Manager as the management node.
+
+After importing or creating a new cluster these pieces of information will be stored in the Uyuni/SUSE Manager database.
+All the described operations will be also exposed in an XMLRPC API.
+
+### Import an existing cluster
+
+The requirement is to have an already provisioned management node for the cluster that the user wants to manage with Uyuni/SUSE Manager.
+It is implied that the management node is already capable of reaching the nodes of the cluster (via passwordless SSH or any other mean requested by the cluster provider manager) and all clustering management stack is already installed in the management node.
+
+When the import action is selected, Uyuni/SUSE Manager will register the management node to Uyuni/SUSE Manager (by replicating what is done under Systems > Bootstrapping).
+Additionally, Uyuni/SUSE Manager asks for the name and the type of cluster that the user wants to register.
```

Comment:
Yes we are, but how can we differentiate at the Uyuni/SUSE Manager level if we want to target a CaaSP or a SES cluster if they share the same management node and the same name?

Reply:
> Yes we are, but how can we differentiate at the Uyuni/SUSE Manager level if we want to target a CaaSP or a SES cluster if they share the same management node and the same name?

I see.

Will we have a "what kind of cluster?" call in the cluster manager API? Maybe that would help:
1. Connect to management node
2. Ask "how many cluster managers are installed on you?"
3. For each cluster manager, ask "what kind of cluster do you manage?"
4. For each cluster manager, ask "what are the names and nodes of each cluster you can manage?"

(or maybe that's already what this RFC is doing and I have not understood it)

Reply:
What you are proposing is inverting the control: the management node has the final say in cluster types because we are asking it "what cluster manager providers are installed on you?".

I think the general philosophy we want to purse with Uyuni/SUSE Manager is to centralize everything into Uyuni/SUSE Manager. In this regard, the question you are suggesting is posed to Uyuni/SUSE Manager when it presents the cluster types available for selection: "what cluster can *I* deploy and manage?".

Does it make sense to you too?

Reply:
A bit philosophical, but Salt's paradigm is more on @paususe's side: As long as you pass the right parameters used by the (virtual or actual) module, the management server (Salt Master) doesn't have to know anything specific about the modules.

Reply:
> What you are proposing is inverting the control: the management node has the final say in cluster types because we are asking it "what cluster manager providers are installed on you?".
> 
> I think the general philosophy we want to purse with Uyuni/SUSE Manager is to centralize everything into Uyuni/SUSE Manager. In this regard, the question you are suggesting is posed to Uyuni/SUSE Manager when it presents the cluster types available for selection: "what cluster can _I_ deploy and manage?".
> 
> Does it make sense to you too?

In order for Uyuni to be able to deploy a cluster of type X, some cluster manager plugin needs to be installed to Uyuni Server, so that list can be populated already.

Then, next thing is what kind of clusters can be managed. That has two requirements:
1. Uyuni Server has the right cluster manager plugin installed. If the cluster already existed but it was not installed by Uyuni, it may very well be the case this cluster manager plugin is not installed. E. g. someone installs a CaaSP cluster manually.
2. There is a cluster manager node. Then we go to the node an ask him "Sir, what kind of clusters can you manage", and the cluster manager node answer "Oh, I know very well about something called CaaSP and SES, and I can offer you clusterA, clusterB and clusterC of type CaaSP, and clusterD and cluster E of type SES. But you know what? That 'CaaSP' and 'SES' are really text strings because in the end Mr Uyuni, you only need to call the methods your base cluster manager interface defined: list_nodes, update_node, etc. So you call them like caasp.clusterA.list_nodes and I will give you something back".

Is this it?

Reply:
I will try to explain myself better.

We are in the "importing a cluster" scenario, where the user already has a cluster already deployed and working:

- Uyuni/SUSE Manager knows that it can manage CaaSP, SES, and SHAP clusters (because the cluster provider manager is installed into the same Uyuni/SUSE Manager server)
- Now the user wants to import the cluster: Uyuni/SUSE Manager asks for the cluster name and the cluster type (from the list gathered above).
- Uyuni/SUSE Manager will ask the management node and proceed to register it to Uyuni/SUSE Manager (if not already registered to Uyuni/SUSE Manager).
- Uyuni/SUSE Manager makes sure that the cluster management stack for the imported cluster is installed in the management node.

Every subsequent action on the cluster will be as described: `salt <mgmt node> <cluster provider mgr>.list_nodes` (and the required parameters) - nothing specific is required

I added this clarification example in this section.

Reply:
But what if the user selects the wrong cluster type? Or makes a type while typing the cluster name? E. g. I want to import a CaaSP cluster but I mistakenly say it's a SES cluster. I guess it will fail.

Those mistakes are something we can easily prevent because we can ask the cluster manager what types of cluster it knows about, and the cluster names for each type of cluster, and then populate that.

Reply:
I think it will be an implementation detail and it will be discussed in the implementation phase. Sorry, I do not think it should be discussed at this level of detail into an RFC.